pkg/controller/release/release_controller.go

```diff
@@ -454,41 +400,122 @@ func (c *Controller) applicationReleases(rel *shipper.Release) ([]*shipper.Relea
+	rel := relinfo.release.DeepCopy()
+
+	releases, err := c.applicationReleases(rel)
 	if err != nil {
-		return false, nil, nil, err
+		return nil, nil, err
 	}
-	prev, succ, err := releaseutil.GetSiblingReleases(relinfo.release, releases)
+	prev, succ, err := releaseutil.GetSiblingReleases(rel, releases)
 	if err != nil {
-		return false, nil, nil, err
+		return nil, nil, err
 	}
 
 	var relinfoPrev, relinfoSucc *releaseInfo
 	if prev != nil {
 		relinfoPrev, err = c.buildReleaseInfo(prev)
 		if err != nil {
-			return false, nil, nil, err
+			return nil, nil, err
 		}
 	}
 	if succ != nil {
 		relinfoSucc, err = c.buildReleaseInfo(succ)
 		if err != nil {
-			return false, nil, nil, err
+			return nil, nil, err
 		}
 	}
 
-	executor := NewStrategyExecutor(relinfoPrev, relinfo, relinfoSucc)
+	isHead := succ == nil
+	var strategy *shipper.RolloutStrategy
+	var targetStep int32
+	// A head release uses it's local spec-defined strategy, any other release
+	// follows it's successor state, therefore looking into the forecoming spec.
+	if isHead {
+		strategy = rel.Spec.Environment.Strategy
+		targetStep = rel.Spec.TargetStep
+	} else {
+		strategy = succ.Spec.Environment.Strategy
+		targetStep = succ.Spec.TargetStep
+	}
+
+	// Looks like a malformed input. Informing about a problem and bailing out.
+	if targetStep >= int32(len(strategy.Steps)) {
+		err := fmt.Errorf("no step %d in strategy for Release %q",
+			targetStep, controller.MetaKey(rel))
+		return nil, nil, shippererrors.NewUnrecoverableError(err)
+	}
+
+	executor := NewStrategyExecutor(strategy, targetStep)
 
-	complete, patches, trans, err := executor.Execute()
+	complete, patches, trans := executor.Execute(relinfoPrev, relinfo, relinfoSucc)
 
 	if len(patches) == 0 {
-		klog.V(4).Infof("Strategy verified for release %q, nothing to patch", controller.MetaKey(relinfo.release))
+		klog.V(4).Infof("Strategy verified for release %q, nothing to patch", controller.MetaKey(rel))
 	} else {
-		klog.V(4).Infof("Strategy has been executed for release %q, applying patches", controller.MetaKey(relinfo.release))
+		klog.V(4).Infof("Strategy has been executed for release %q, applying patches", controller.MetaKey(rel))
+	}
+
+	condition := releaseutil.NewReleaseCondition(
+		shipper.ReleaseConditionTypeStrategyExecuted,
+		corev1.ConditionTrue,
+		"",
+		"",
+	)
+	diff.Append(releaseutil.SetReleaseCondition(&rel.Status, *condition))
+
+	isLastStep := int(targetStep) == len(strategy.Steps)-1
+	prevStep := rel.Status.AchievedStep
+
+	if complete {
+		var achievedStep int32
+		var achievedStepName string
+		if isHead {
+			achievedStep = targetStep
+			achievedStepName = strategy.Steps[achievedStep].Name
+		} else {
+			achievedStep = int32(len(rel.Spec.Environment.Strategy.Steps)) - 1
+			achievedStepName = rel.Spec.Environment.Strategy.Steps[achievedStep].Name
+
+		}
+		if prevStep == nil || achievedStep != prevStep.Step {
```

Comment:
I have a small concern: we are trusting the pipeline to not be `complete` when the target step was not achieved, so if we arrived here, it means 100% target step was achieved and achievedStep == targetStep?

Reply:
It's quite confusing indeed. Achieved step is the one we're reporting. targetStep might be different as it's the one that comes from the successor (if any) strategy. If the current release is not head, here in this very place we simply ensure that it's state reports the last step in it's own strategy, no in it's successor which is being used as a desired state.